File: module-9-self-improve/solutions/python/improve.py
"""Module 9: Self-improvement meta-agent."""
import os
import sys
import json

sys.path.insert(0, os.path.dirname(__file__))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-7-evals", "solutions", "python"))

from dotenv import load_dotenv
import anthropic
from eval_types import EvalReport

load_dotenv()

client = anthropic.Anthropic()


def analyze_and_improve(report: EvalReport, agent_source_dir: str) -> list[dict]:
    prompt_path = os.path.join(agent_source_dir, "module-5-coding-agent", "solutions", "python", "prompt.py")
    with open(prompt_path) as f:
        prompt_source = f.read()

    failures = []
    for task_result in report.tasks:
        for trial in task_result.trials:
            if not trial.grade.passed:
                failures.append({
                    "task": trial.task_id,
                    "explanation": trial.grade.explanation,
                    "turns": len(trial.transcript.turns),
                    "tool_calls": [
                        tc["name"]
                        for turn in trial.transcript.turns
                        for tc in (turn.tool_calls or [])
                    ],
                })

    if not failures:
        print("No failures to analyze — agent is performing well!")
        return []

    print(f"Analyzing {len(failures)} failures across {len(report.tasks)} tasks...")

    failure_text = "\n\n".join(
        f"- Task: {f['task']}\n  Explanation: {f['explanation']}\n  Turns used: {f['turns']}\n  Tools used: {', '.join(f['tool_calls'])}"
        for f in failures
    )

    response = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=4096,
        messages=[{
            "role": "user",
            "content": f"""You are an agent improvement specialist. Analyze these eval failures and propose improvements to the agent's system prompt.

## Current System Prompt
```python
{prompt_source}
```

## Eval Scorecard
Overall pass rate: {report.overall_pass_rate * 100:.1f}%

## Failure Details
{failure_text}

## Instructions
1. Identify recurring failure patterns
2. For each pattern, propose a specific addition to the system prompt
3. Return ONLY valid JSON array:
[{{"description": "what this fixes", "addition": "text to add to the prompt rules"}}]

Be specific and actionable. Don't add vague advice.""",
        }],
    )

    text = response.content[0]
    if text.type != "text":
        return []

    try:
        suggestions = json.loads(text.text)
        improvements = []
        updated_prompt = prompt_source

        for suggestion in suggestions:
            import re
            match = re.search(r'## Rules\n([\s\S]*?)(?=\n\n## |"""\s*$)', updated_prompt)
            if match:
                rule_lines = match.group(1).strip().split("\n")
                next_num = len(rule_lines) + 1
                new_rule = f"{next_num}. {suggestion['addition']}"
                new_rules = match.group(1).rstrip() + "\n" + new_rule
                updated_prompt = updated_prompt.replace(match.group(1), new_rules)
                improvements.append({
                    "file": "module-5-coding-agent/solutions/python/prompt.py",
                    "description": suggestion["description"],
                    "old_content": match.group(1),
                    "new_content": new_rules,
                })

        return improvements
    except Exception:
        print("Failed to parse improvement suggestions")
        return []


def apply_improvements(improvements: list[dict], base_dir: str) -> None:
    for imp in improvements:
        file_path = os.path.join(base_dir, imp["file"])
        with open(file_path) as f:
            content = f.read()
        content = content.replace(imp["old_content"], imp["new_content"])
        with open(file_path, "w") as f:
            f.write(content)
        print(f"Applied: {imp['description']}")

File: module-9-self-improve/solutions/improve.ts
/**
 * Module 9: Self-improvement meta-agent
 * Reads eval scorecard + agent source, proposes and applies fixes.
 */
import "dotenv/config";
import * as fs from "fs/promises";
import * as path from "path";
import Anthropic from "@anthropic-ai/sdk";
import type { EvalReport } from "../../module-7-evals/solutions/types.js";

const client = new Anthropic();

export interface Improvement {
  file: string;
  description: string;
  oldContent: string;
  newContent: string;
}

export async function analyzeAndImprove(
  report: EvalReport,
  agentSourceDir: string
): Promise<Improvement[]> {
  // Read agent source files
  const promptPath = path.join(agentSourceDir, "module-5-coding-agent", "prompt.ts");
  const promptSource = await fs.readFile(promptPath, "utf-8");

  // Extract failure patterns from report
  const failures = report.tasks
    .flatMap((t) =>
      t.trials
        .filter((trial) => !trial.grade.passed)
        .map((trial) => ({
          task: trial.taskId,
          explanation: trial.grade.explanation,
          turns: trial.transcript.turns.length,
          toolCalls: trial.transcript.turns
            .flatMap((turn) => turn.toolCalls || [])
            .map((tc) => tc.name),
        }))
    );

  if (failures.length === 0) {
    console.log("No failures to analyze — agent is performing well!");
    return [];
  }

  console.log(`Analyzing ${failures.length} failures across ${report.tasks.length} tasks...`);

  const response = await client.messages.create({
    model: "claude-sonnet-4-20250514",
    max_tokens: 4096,
    messages: [
      {
        role: "user",
        content: `You are an agent improvement specialist. Analyze these eval failures and propose improvements to the agent's system prompt.

## Current System Prompt
\`\`\`typescript
${promptSource}
\`\`\`

## Eval Scorecard
Overall pass rate: ${(report.overallPassRate * 100).toFixed(1)}%

## Failure Details
${failures.map((f) => `- Task: ${f.task}\n  Explanation: ${f.explanation}\n  Turns used: ${f.turns}\n  Tools used: ${f.toolCalls.join(", ")}`).join("\n\n")}

## Instructions
1. Identify recurring failure patterns (e.g., "edits without reading first", "wrong search pattern")
2. For each pattern, propose a specific addition to the system prompt
3. Return ONLY valid JSON array:
[{"description": "what this fixes", "addition": "text to add to the prompt rules"}]

Be specific and actionable. Don't add vague advice.`,
      },
    ],
  });

  const text = response.content[0];
  if (text.type !== "text") return [];

  try {
    const suggestions: { description: string; addition: string }[] = JSON.parse(text.text);

    // Apply suggestions to prompt
    let updatedPrompt = promptSource;
    const improvements: Improvement[] = [];

    for (const suggestion of suggestions) {
      const oldRules = updatedPrompt.match(/## Rules\n([\s\S]*?)(?=\n\n## |`;\s*$)/);
      if (oldRules) {
        const ruleLines = oldRules[1].trim().split("\n");
        const nextNum = ruleLines.length + 1;
        const newRule = `${nextNum}. ${suggestion.addition}`;
        const newRules = oldRules[1].trimEnd() + "\n" + newRule;
        updatedPrompt = updatedPrompt.replace(oldRules[1], newRules);

        improvements.push({
          file: "module-5-coding-agent/prompt.ts",
          description: suggestion.description,
          oldContent: oldRules[1],
          newContent: newRules,
        });
      }
    }

    return improvements;
  } catch {
    console.error("Failed to parse improvement suggestions");
    return [];
  }
}

export async function applyImprovements(
  improvements: Improvement[],
  baseDir: string
): Promise<void> {
  for (const imp of improvements) {
    const filePath = path.join(baseDir, imp.file);
    let content = await fs.readFile(filePath, "utf-8");
    content = content.replace(imp.oldContent, imp.newContent);
    await fs.writeFile(filePath, content, "utf-8");
    console.log(`Applied: ${imp.description}`);
  }
}
